service/tbc/crawler.go

```diff
@@ -124,94 +161,149 @@ func (s *Server) fixupCache(ctx context.Context, b *btcutil.Block, utxos map[tbc
 		}
 
 		blocksProcessed++
 
 		// Try not to overshoot the cache to prevent costly allocations
 		cp := len(utxos) * 100 / s.cfg.MaxCachedTxs
-		if height%10000 == 0 || cp > utxosPercentage || blocksProcessed == 1 {
-			log.Infof("Utxo indexer height: %v utxo cache %v%%", height, cp)
+		if bh.Height%10000 == 0 || cp > utxosPercentage || blocksProcessed == 1 {
+			log.Infof("Utxo indexer: %v utxo cache %v%%", hh, cp)
 		}
 		if cp > utxosPercentage {
 			// Set utxosMax to the largest utxo capacity seen
 			s.cfg.MaxCachedTxs = max(len(utxos), s.cfg.MaxCachedTxs)
+			last = hh
 			// Flush
 			break
 		}
 
-		// If set we may have to exit early
-		if circuitBreaker {
-			if height >= maxHeight-1 {
+		// Exit if we processed the provided end hash
+		if endHash.IsEqual(&hash) {
+			last = hh
+			break
+		}
+
+		// Move to next block
+		height := bh.Height + 1
+		bhs, err := s.db.BlockHeadersByHeight(ctx, height)
+		if err != nil {
+			if errors.Is(err, database.ErrNotFound) {
+				log.Infof("No more blocks at: %v", height)
 				break
 			}
+			return 0, last, fmt.Errorf("block headers by height %v: %w",
+				height, err)
+		}
+		if len(bhs) > 1 {
+			panic("FIXME handle multiple block headers")
 		}
+		// Verify it connects to parent
+		if !hash.IsEqual(bhs[0].ParentHash()) {
+			return 0, last, fmt.Errorf("%v does not connect to: %v",
+				bhs[0], hash)
+		}
+		hh.Hash = *bhs[0].BlockHash()
+		hh.Height = bhs[0].Height
 	}
 
-	return blocksProcessed, nil
+	return blocksProcessed, last, nil
 }
 
-func (s *Server) UtxoIndexer(ctx context.Context, height, count uint64) error {
+func (s *Server) UtxoIndexer(ctx context.Context, endHash *chainhash.Hash) error {
 	log.Tracef("UtxoIndexer")
 	defer log.Tracef("UtxoIndexer exit")
 
-	var maxHeight uint64
-	circuitBreaker := false
-	if count != 0 {
-		circuitBreaker = true
-		maxHeight = height + count
+	// Verify exit condition hash
+	if endHash == nil {
+		return fmt.Errorf("must provide an end hash")
+	}
+	_, endHeight, err := s.BlockHeaderByHash(ctx, endHash)
+	if err != nil {
+		return fmt.Errorf("blockheader hash: %w", err)
+	}
+
+	// Verify start point is not after the end point
+	utxoHH, err := s.UtxoIndexHash(ctx)
+	if err != nil {
+		if !errors.Is(err, database.ErrNotFound) {
+			return fmt.Errorf("utxo indexer : %w", err)
+		}
+		utxoHH = &HashHeight{
+			Hash:   *s.chainParams.GenesisHash,
+			Height: 0,
+		}
 	}
+	// XXX we need training wheels here. We can't blind accept the end
+	// without asserting if it is either ihigher in the chain or is a
+	// forced for.
+	// XXX check cumulative? check fork?
 
 	// Allocate here so that we don't waste space when not indexing.
 	utxos := make(map[tbcd.Outpoint]tbcd.CacheOutput, s.cfg.MaxCachedTxs)
 	defer clear(utxos)
 
-	log.Infof("Start indexing UTxos at height %v count %v", height, count)
+	log.Infof("Start indexing UTxos at hash %v height %v", utxoHH.Hash, utxoHH.Height)
+	log.Infof("End indexing UTxos at hash %v height %v", endHash, endHeight)
```

Comment:
Should these logs be at the `info` level? If they are primarily for debugging, I think the `debug` level would be better.

Reply:
I have actually gone back and forth on this. If we don't print it we have absolutely no idea how long it'll run or how much needs to be done. In principle I agree with you but we need a start and end indication the way the code works now.

Reply:
Ah okay, that's fine. I just don't want to end up with random unneeded Info logs :)